File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {

    int help(int idx , int[] dp, int[] nums, int st){
      if(idx<st) return 0;
      if(idx == st){
        return nums[idx];
      }

      if(dp[idx] != -1){
        return dp[idx];
      }

      int take = 0, ntake = 0;
      
     take = nums[idx] + help(idx-2,dp,nums,st);
      ntake = help(idx-1,dp,nums,st);

      return dp[idx] = Math.max(take,ntake);
    }
    public int rob(int[] nums) {
        if(nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        int[] dp = new int[nums.length];
        Arrays.fill(dp,-1);
         int n = nums.length-1;
        int a = help(n,dp,nums,1);

         Arrays.fill(dp,-1);
        
        int b = help(n-1,dp,nums,0);
        return Math.max(a,b);
       
       
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int helper(int idx, int start, vector<int>& nums, vector<int>& dp) {
        if (idx < start) return 0;
        if (idx == start) return nums[start];

        if (dp[idx] != -1) return dp[idx];

        int pick = nums[idx] + helper(idx - 2, start, nums, dp);
        int notPick = helper(idx - 1, start, nums, dp);

        return dp[idx] = max(pick, notPick);
    }

    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];

        // Case 1: exclude last house
        vector<int> dp1(n, -1);
        int case1 = helper(n - 2, 0, nums, dp1);

        // Case 2: exclude first house
        vector<int> dp2(n, -1);
        int case2 = helper(n - 1, 1, nums, dp2);

        return max(case1, case2);
    }
};
